pkg/credentials/service_account_credentials.go

```diff
@@ -63,6 +71,87 @@ func NewCredentialBulder(client client.Client, config *v1.ConfigMap) *Credential
 	}
 }
 
+func (c *CredentialBuilder) CreateStorageSpecSecretEnvs(namespace string, storageKey string,
+	storageSecretName string, overrideParams map[string]string, container *v1.Container) error {
+
+	stype, ok := overrideParams["type"]
+	if ok && stype != "s3" {
+		return errors.New("only S3 type storage is currently supported with storage spec")
+	}
+	if u, err := url.Parse(container.Args[0]); err == nil {
+		if stype = u.Scheme; stype != "s3" {
+			return errors.New("only S3 type storage is currently supported with storage spec")
+		}
+	}
+
+	secret := &v1.Secret{}
+	var storageData []byte
+	if err := c.client.Get(context.TODO(),
+		types.NamespacedName{Name: storageSecretName, Namespace: namespace}, secret); err == nil {
+		if storageKey != "" {
+			if storageData = secret.Data[storageKey]; storageData == nil {
+				return fmt.Errorf("specified storage key %s not found in storage secret %s",
+					storageKey, storageSecretName)
+			}
+		} else {
+			if stype == "" {
+				storageKey = DefaultStorageSecretKey
+			} else {
+				storageKey = fmt.Sprintf("%s_%s", DefaultStorageSecretKey, stype)
+			}
+			// It's ok for the entry not to be found in the default/fallback cases
+			storageData = secret.Data[storageKey]
+		}
+	} else if storageKey != "" || !apierr.IsNotFound(err) { // Don't fail if not found and no storage key was specified
+		return fmt.Errorf("can't read storage secret %s: %w", storageSecretName, err)
+	}
+
+	if storageData != nil {
+		if stype == "" {
+			var storageDataJson map[string]string
+			if err := json.Unmarshal(storageData, &storageDataJson); err != nil {
+				return fmt.Errorf("invalid json encountered in key %s of storage secret %s: %w",
+					storageKey, storageSecretName, err)
+			}
+			if stype, ok = storageDataJson["type"]; ok && stype != "s3" {
+				return errors.New("only S3 type storage is currently supported with storage spec")
+			}
+		}
+		// Pass storage config json as SecretKeyRef env var
+		container.Env = append(container.Env, v1.EnvVar{
+			Name: StorageConfigEnvKey,
+			ValueFrom: &v1.EnvVarSource{
+				SecretKeyRef: &v1.SecretKeySelector{
+					LocalObjectReference: v1.LocalObjectReference{
+						Name: storageSecretName,
+					},
+					Key: storageKey,
+				},
+			},
+		})
+	}
+
+	if stype == "" {
+		return errors.New("unable to determine storage type")
+	}
+
+	if strings.HasPrefix(container.Args[0], UriSchemePlaceholder+"://") {
+		container.Args[0] = stype + container.Args[0][20:]
```

Comment:
what's `container.Args[0][20:]` here?

Reply:
This is the StorageURI without the storage type prefix, so we can replace the UriSchemePlaceholder  `<scheme-placeholder>` to stype `s3`. Sorry let me change back from `20` to `len(UriSchemePlaceholder)`